.github/workflows/ci.yml

```diff
@@ -2,20 +2,26 @@
 name: ci
 on:
   push:
-#    branches:
-#      - master
+    branches:
+      - master
   pull_request:
     branches:
       - master
 jobs:
   ci:
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      max-parallel: 6
+      matrix:
+        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11", "pypy-3.9"]
     steps:
     - uses: actions/checkout@v3
     - uses: actions/setup-python@v4
       with:
-        python-version: 3.x
+        python-version: ${{ matrix.python-version }}
     - run: pip install --upgrade pip
-    - run: pip install pytest ruff .
-    - run: ruff --format=github --ignore=E501 --target-version=py37 .
+    - if: ${{ matrix.python-version }} == "3.11"
+      run: pip install ruff && ruff --format=github --ignore=E501 --target-version=py37 .
```

Comment:
Is ruff still necessary? It seems tests are executed with pytests below anyway?

Reply:
Ruff is a linter and pytest runs the test suite.
Pytest does not run the ruff tests.
To prove this, add a ruff error and only the Python 3.11 run will fail.

Reply:
Cool thanks for explanation. It's my ignorance for what Ruff does.